File: frontend/src/helpdesk/types/ticket.ts
import type { ContactIn, ContactSummary } from '@/helpdesk/types/contact'

export type TicketStatus = 'open' | 'pending' | 'resolved' | 'closed'
export type TicketPriority = 'low' | 'normal' | 'high' | 'urgent'
export type TicketChannel = 'agent' | 'widget'

export interface UserSummary {
  id: number
  name: string
}

export interface TicketOut {
  id: number
  organization_id: number
  number: number
  subject: string
  status: TicketStatus
  priority: TicketPriority
  channel: TicketChannel
  contact: ContactSummary
  assignee: UserSummary | null
  first_response_at: string | null
  resolved_at: string | null
  closed_at: string | null
  last_message_at: string
  created_at: string
  updated_at: string
}

export interface TicketMessageOut {
  id: number
  ticket_id: number
  author_type: 'agent' | 'contact'
  author_user_id: number | null
  author_contact_id: number | null
  author_name: string | null
  body: string
  is_internal: boolean
  created_at: string
}

export interface TicketDetailOut extends TicketOut {
  messages: TicketMessageOut[]
}

export interface TicketIn {
  subject: string
  body: string
  priority?: TicketPriority
  /** Exactly one of `contact_id` or `contact` (matched by email, created when new). */
  contact_id?: number
  contact?: ContactIn
  assignee_id?: number | null
}

export interface TicketPatch {
  subject?: string
  status?: TicketStatus
  priority?: TicketPriority
}

export interface TicketMessageIn {
  body: string
  is_internal?: boolean
}

File: backend/src/helpdesk/schemas/ticket.py
from datetime import datetime
from typing import Annotated, Self

from pydantic import BaseModel, ConfigDict, Field, model_validator

from src.helpdesk.enums import (
    MessageAuthorType,
    TicketChannel,
    TicketPriority,
    TicketStatus,
)
from src.helpdesk.schemas.contact import ContactIn, ContactSummary
from src.platform.schemas.common import Timestamp

TicketSubject = Annotated[str, Field(min_length=1, max_length=255)]
MessageBody = Annotated[str, Field(min_length=1, max_length=20_000)]


class UserSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str


class TicketIn(BaseModel):
    subject: TicketSubject
    body: MessageBody
    priority: TicketPriority = TicketPriority.NORMAL
    # Either an existing contact, or the details of the customer who got in
    # touch - matched to an existing contact by email, created when new.
    contact_id: int | None = None
    contact: ContactIn | None = None
    assignee_id: int | None = None

    @model_validator(mode="after")
    def _exactly_one_contact(self) -> Self:
        if (self.contact_id is None) == (self.contact is None):
            raise ValueError("Provide exactly one of contact_id or contact")
        return self


class TicketPatch(BaseModel):
    subject: TicketSubject | None = None
    status: TicketStatus | None = None
    priority: TicketPriority | None = None


class TicketAssigneeIn(BaseModel):
    # null returns the ticket to the unassigned queue.
    assignee_id: int | None


class TicketMessageIn(BaseModel):
    body: MessageBody
    is_internal: bool = False


class TicketMessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    ticket_id: int
    author_type: MessageAuthorType
    author_user_id: int | None
    author_contact_id: int | None
    author_name: str | None
    body: str
    is_internal: bool
    created_at: datetime


class TicketOut(Timestamp):
    model_config = ConfigDict(from_attributes=True)

    id: int
    organization_id: int
    number: int
    subject: str
    status: TicketStatus
    priority: TicketPriority
    channel: TicketChannel
    contact: ContactSummary
    assignee: UserSummary | None
    first_response_at: datetime | None
    resolved_at: datetime | None
    closed_at: datetime | None
    last_message_at: datetime


class TicketDetailOut(TicketOut):
    messages: list[TicketMessageOut]
